tinygrad/tensor.py

```diff
@@ -265,78 +265,66 @@ def pad(self, arg: Tuple[Tuple[int, int], ...], value:float=0) -> Tensor:
-            if tensor_found[i_][0] > i: sub[i_] -= 1
-      else: # s is None
-        final_shape.append(1)
-    ret = sliced_tensor.reshape(tuple(final_shape))  # Reshape
-    if tensor_found: # Fancy/tensor indexing
-      for i,s in enumerate(sub): tensor_found[i] = (tensor_found[i][0]+s, tensor_found[i][1])
-      dim = [i[0] for i in tensor_found]
-      idx = [i[1].sign().contiguous().__neg__().contiguous().relu() * ret.shape[i[0]] + i[1] for i in tensor_found] # TODO first contiguous fixes torch+cpu_only CI, but it causes llvm to fail. Second one fixes llvm
+        if isinstance(s, (slice, Tensor)): final_shape.append(dim_shape)
+        elif tensors: # s is int
+          for i_ in range(len(orig_dim)):
+            if orig_dim[i_] > i: dim[i_] -= 1 # trim dims associated with tensors after dim is collapsed
+    ret = sliced_tensor.reshape(tuple(final_shape))
+    # Fancy/tensor indexing
+    if tensors:
+      # normalize idx
+      idx = [t.sign().contiguous().__neg__().contiguous().relu() * ret.shape[d] + t for d,t in zip(dim, tensors)] # TODO first contiguous fixes torch+cpu_only CI, but it causes llvm to fail. Second one fixes llvm
+      # compute sum_dim, arange, and idx
       max_dim = max(i.ndim for i in idx)
-      idx = [i.reshape(*[1]*(max_dim-i.ndim), *i.shape) for i in idx]
-      sum_dim = [d+max_dim-n for n,d in enumerate(dim)]
-      new_idx = idx[0].reshape(*[1]*dim[0], 1,*idx[0].shape, *[1]*(ret.ndim-dim[0]-1))
-      arange = Tensor.arange(ret.shape[dim[0]], dtype=dtypes.int32, requires_grad=False, device=self.device).reshape(*[1]*dim[0], ret.shape[dim[0]], *[1]*idx[0].ndim, *[1]*(ret.ndim-dim[0]-1))
-      ret = (ret.reshape(*ret.shape[:dim[0]+1], *[1]*idx[0].ndim, *ret.shape[dim[0]+1:]) * (arange == new_idx)).sum(dim[0])
-      for idx_,d in zip(idx[1:],sum_dim[1:]):
-        new_idx = idx_.reshape(*[1]*dim[0], *idx_.shape, *[1]*(ret.ndim-dim[0]-idx_.ndim))
-        arange = Tensor.arange(ret.shape[d], dtype=dtypes.int32, requires_grad=False, device=self.device).reshape(*[1]*(d), ret.shape[d], *[1]*(ret.ndim-d-1))
-        ret = ((new_idx == arange) * ret).sum(d)
-      if dim[0] != 0 and dim != list(range(dim[0], dim[-1]+1)) and len(dim) != 1: # special permute case
-        order = list(range(ret.ndim))
-        order = order[dim[0]:dim[0]+idx[0].ndim] + order[:dim[0]] + order[dim[0]+idx[0].ndim:]
-        ret = ret.permute(order=order)
+      sum_dim = [d if n==0 else d+max_dim-n for n,d in enumerate(dim)]
+      arange = [Tensor.arange(ret.shape[d], dtype=dtypes.int32, requires_grad=False, device=self.device).reshape(*[1]*sd, ret.shape[d], *[1]*(((ret.ndim + max_dim) if n==0 else (ret.ndim + max_dim - n)) - sd - 1)) for n,(sd,d) in enumerate(zip(sum_dim, dim))]
+      idx = [i.reshape(*[1]*sum_dim[0], *((1, *[1]*(max_dim-i.ndim), *i.shape, *[1]*(ret.ndim - sd - 1)) if n == 0 else (*[1]*(max_dim-i.ndim), *i.shape, *[1]*(ret.ndim - sum_dim[0] - n)))) for n,(sd,i) in enumerate(zip(sum_dim, idx))]
```

Comment:
this is pretty unreadable, can you split this line up a bit?

Reply:
Really happy with the new arange change. 
Current arange is readable and I separated idx 1 liner to 3 lines. Probably more readable